core/state/statedb.go

```diff
@@ -680,46 +712,40 @@ func (self *StateDB) Prepare(thash, bhash common.Hash, ti int) {
 }
 
 func (s *StateDB) clearJournalAndRefund() {
-	s.journal = newJournal()
-	s.validRevisions = s.validRevisions[:0]
-	s.refund = 0
+	if len(s.journal.entries) > 0 {
+		s.journal = newJournal()
+		s.validRevisions = s.validRevisions[:0]
```

Comment:
I'm not 100% sure that this is correct. Are there cases where the `journal.entries` is empty, but `s.validRevisions` is non-empty? If so, this statement should be moved out of the if-clause, and same with the statement below

Reply:
Refunds are tracked by journal entries (AddRefund/SubRefud both add to the journal), so if the refound counter is non zero, the journal must have something in it.

As for the validRevisions, I guess doing nested CALLs without anything else would create snapshots without journal entries that would not get cleaned up. Seems easier/safer to move out of the if and make sure we clean it up. The worst case I can imagine is a tiny memory leak that gets fixed the moment a subsequent tx touches the journal, but still, better to have correct logic.

Reply:
PTAL